SI-3368 CDATA gets a Node

som-snytt opened the pull request:
XML Parser uses `scala.xml.PCData`.

A compiler flag `-Yxml:coalescing`, analogous to
`DocumentBuilderFactory.setCoalescing`, turns `PCData`
nodes into `Text` nodes and coalesces sibling text nodes.

This change also fixes parse errors such as rejecting a
sequence of CDATA sections.

A sequence of "top level" nodes are not coalesced.

```
scala> <a><b/>start<![CDATA[hi & bye]]><c/>world<d/>stuff<![CDATA[red & black]]></a>
res0: scala.xml.Elem = <a><b/>start<![CDATA[hi & bye]]><c/>world<d/>stuff<![CDATA[red & black]]></a>

scala> :replay -Yxml:coalescing
Replaying: <a><b/>start<![CDATA[hi & bye]]><c/>world<d/>stuff<![CDATA[red & black]]></a>
res0: scala.xml.Elem = <a><b/>starthi &amp; bye<c/>world<d/>stuffred &amp; black</a>
```

som-snytt committed:
SI-3368 CDATA gets a Node

XML Parser uses `scala.xml.PCData`.

A compiler flag `-Yxml:coalescing`, analogous to
`DocumentBuilderFactory.setCoalescing`, turns `PCData`
nodes into `Text` nodes and coalesces sibling text nodes.

This change also fixes parse errors such as rejecting a
sequence of CDATA sections.

A sequence of "top level" nodes are not coalesced.

```
scala> <a><b/>start<![CDATA[hi & bye]]><c/>world<d/>stuff<![CDATA[red & black]]></a>
res0: scala.xml.Elem = <a><b/>start<![CDATA[hi & bye]]><c/>world<d/>stuff<![CDATA[red & black]]></a>

scala> :replay -Yxml:coalescing
Replaying: <a><b/>start<![CDATA[hi & bye]]><c/>world<d/>stuff<![CDATA[red & black]]></a>
res0: scala.xml.Elem = <a><b/>starthi &amp; bye<c/>world<d/>stuffred &amp; black</a>
```

som-snytt committed:
SI-5699 Use ParserTest

Update the test slightly to use the rig it inspired.

mbeckerle commented:
Thank you for fixing this bug. I wanted to bring to your attention another issue that is symmetric with this one on the print side of XML. Really to read stuff in preserving CDATA, and print it out as well requires fixing that one as well. https://issues.scala-lang.org/browse/SI-4303. I voted for that bug, but thought it worth mentioning here.

som-snytt commented:
@mbeckerle Thx, I'll take a look. I don't use xml anymore, knock on wood, but a few low-hanging fruit are in scope. Soon all your bugs are belong to xml-interpolator.

gkossakowski commented:
Hi @som-snytt!

Thanks for looking into this issue. I was assigned to review this PR. I have a question about the `-Yxml:coalescin` option. Is it meant just as a fallback in Scala 2.11 just in case somebody relies on old behavior? Do you want the new behavior to be default in 2.12?

som-snytt commented:
@gkossakowski Thanks for looking. My eyes on this are now fresh, even virginal.

Javadoc says DocumentBuilderFactory.setCoalescing defaults to false, so off by default is appropriate.

Why did I write "transitional in 2.11" on the option? Not sure. I may have meant "Bruce Jenner transitioning" in the sense of going from `-Y` to `-X`, or acquiring more behaviors. I may have meant that it should be configurable on the provider side: scala-xml and friends should implement these behaviors, hence `-Y`.

Or, I may have meant `-Y` is ephemeral and support may disappear when xml literals are dropped in favor of xml interpolators. I probably meant that.

gkossakowski commented:
I see. Let's rename it to `-X` because we'll definitely keep it for longer than one release and it feels like rather standard option even if for a bit niche feature (xml parsing).

som-snytt committed:
SI-3368 Promote xml option to -Xxml

As long as Scala does XML literals, there is probably parsing
behavior worth configuring.

Therefore, the umbrella option is promoted to `-Xxml`.

It was tempting to make it `-XML`, but we resisted.

som-snytt commented:
Rebased and renamed.

adriaanm commented:
Could you open a new PR for this one, @som-snytt? Scabot eagerly awaits the rebased version.

som-snytt commented:
See https://github.com/scala/scala/pull/4306.

Is Scabot pronounced the French way?
